src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1612,46 +1588,77 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
+                    if (!DoesParamArgRankMatchInner(fep, p, t, argIndex)) return false;
                 }
-                else if (args[argIndex] is IdentifierNode idn)
+                else if (currentArg is IdentifierNode idn)
                 {
                     if (idn.ReplicationGuides.Count > 0) return false;
 
                     var t = variables[idn.Value].Item2;
                     if (t == typeof(object)) return false;
 
-                    if (!p.Equals(t))
+                    if(!DoesParamArgRankMatchInner(fep,p, t,argIndex)) return false;
+                }
+                else if (currentArg is RangeExprNode rgnNode)
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamArgRankMatchInner(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        private static bool DoesParamArgRankMatchInner(ProtoCore.CLRFunctionEndPoint fep,Type p, Type t, int paramIndex)
         {
+            if (!p.Equals(t))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (p.IsValueType && t.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(fep,t,paramIndex);
+                var paramRank = GetRank(fep,p, paramIndex);
+
+                //if the param is an arbitrary rank array and
+                //arg is some array type, replication should not occur.
+                if (paramRank == -1 && argRank > 0)
+                {
+                    return true;
+                }
+                if (argRank != paramRank) return false;
+
+                // If both have rank 0, it could also be because their type info is ambiguous.
+                if (argRank == 0 && paramRank == 0) return false;
+            }
+            return true;
+        }
+
+        private static int GetRank(ProtoCore.CLRFunctionEndPoint fep,Type type,int paramIndex)
+        {
+            if(type == typeof(IList)){
+                return -1;
+            }
+            if(type == typeof(object) && fep.ParamAttributes[paramIndex].Any(attr => attr is ArbitraryDimensionArrayImportAttribute))
+            {
+                return -1;
+            }
```

Comment:
I believe a parameter of type `IEnumerable<T>`  with `ArbitraryDimensionImportAttribute` results in an arbitrary dimension array of type `T`, in which case the `type == typeof(object)` check doesn't seem quite right, does it?

Reply:
Hi @aparajit-pratap, this check is specifically for a test that was calling `List.Contains` - this method and some of the other core list methods accept object parameters marked with this attribute:
https://github.com/DynamoDS/Dynamo/blob/c9a23ab9c9e4110ac798b33a9d4afad183718034/src/Libraries/CoreNodes/List.cs#L51

Reply:
I guess my question is how do we deal with parameters like this:
```
public void foo([ArbitraryDimensionArrayImport] IEnumerable<int> bar) { ... }
```
that are passed an `int` or list of any rank of `int`s as argument.

Reply:
@aparajit-pratap I need to test it, but I agree it's a concern for example - `FirstItem(list:var[]..[]` should work with a list or a single item, I'm not sure what happens today though, I can add a test for that case at the minimum.

Reply:
ah @aparajit-pratap to clarify, are you asserting that because `int` is a `valueType` that my check with `object` is actually too strict?

Now that I think about it - it does make sense that basically any use of this attribute on a parameter means it cannot replicate - EXCEPT in cases where there are replicate guides that force an iteration - does that sound right?

Reply:
1. I believe there are 2 cases where ArbitraryDimension attribute applies, where it converts from:
a. `IList` to `var[]..[]` - I think you've taken care of this
b. `IEnumerable<T>` to `T[]..[]` - my question was about this case, how would you handle it as I don't think checking for `object` type would work here
2. I'm unsure what happens when it is applied to `object`. If it does, your check for the type being equal to `object` makes sense.
3. I agree, it would be good to test this logic with a single item arg passed to a parameter with the attribute

Reply:
It looks like the `ArbitraryDimension` attribute applies to `object` as well, so you're right about checking for object type. Scratch (2) above.